File: elixir/test/clone_entity_test.exs
# Clone entity test (offline, mock transport)

defmodule BluefinShieldconexMgmt.CloneEntityTest do
  use ExUnit.Case

  alias Voxgig.Struct, as: S
  alias BluefinShieldconexMgmt.Helpers, as: H
  alias BluefinShieldconexMgmt.Json

  defp fixture do
    Json.parse(File.read!("../.sdk/test/entity/clone/CloneTestData.json"))
  end

  defp mk_sdk do
    existing = H.or_(S.getpath(fixture(), "existing"), S.jm([]))
    BluefinShieldconexMgmt.test(S.jm(["entity", existing]))
  end

  defp first_id do
    existing = H.or_(S.getpath(fixture(), "existing.clone"), S.jm([]))
    keys = S.keysof(existing)
    if keys == [], do: nil, else: hd(keys)
  end

  test "should create instance" do
    sdk = BluefinShieldconexMgmt.test()
    ent = BluefinShieldconexMgmt.clone(sdk)
    assert ent != nil
  end

  test "should create then read back" do
    sdk = BluefinShieldconexMgmt.test(S.jm(["entity", S.jm(["clone", S.jm([])])]))
    ent = BluefinShieldconexMgmt.clone(sdk)
    made = BluefinShieldconexMgmt.Entity.Clone.create(ent, S.jm(["name", "test-create"]))
    assert S.ismap(made)
    assert S.getprop(made, "id") != nil
  end
end

File: scala/sdktest/CloneEntityTest.scala
// Generated basic-flow test for the clone entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped CloneTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefinshieldconexmgmtsdk.core.{Helpers, BluefinShieldconexMgmtSDK}
import voxgig.bluefinshieldconexmgmtsdk.utility.struct.Struct

object CloneEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("clone.instance") {
      val testsdk = BluefinShieldconexMgmtSDK.testSDK()
      val ent = testsdk.clone(null)
      rep.check("clone.instance", ent != null, "expected non-null clone entity")
    }

    rep.scope("clone.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/clone/CloneTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinShieldconexMgmtSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("clone01", "CLONE01")
      idmap.put("clone02", "CLONE02")
      idmap.put("clone03", "CLONE03")
      idmap.put("template01", "TEMPLATE01")
      idmap.put("template02", "TEMPLATE02")
      idmap.put("template03", "TEMPLATE03")
      val now = System.currentTimeMillis()

      // CREATE
      val cloneRef01Ent = client.clone(null)
      var cloneRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.clone"), "clone_ref01"))
      cloneRef01Data.put("template_id", idmap.get("template01"))
      val cloneRef01DataResult = cloneRef01Ent.create(cloneRef01Data, null)
      cloneRef01Data = Helpers.toMapAny(cloneRef01DataResult)
      rep.check("clone.create.map", cloneRef01Data != null, "expected create result to be a map")
      rep.check("clone.create.id", cloneRef01Data != null && cloneRef01Data.get("id") != null, "expected created entity to have an id")
    }
  }
}
